Reimplemented Paths framework in C++

johnhaddon opened the pull request:
This is the "SceneHierarchy rewrite" requested in #852. In reality, the SceneHierarchy code is barely touched, and it's the underlying Path and PathListingWidget classes which are rewritten in C++ for improved performance.

Paths can still be implemented in Python (and several are still in Gaffer), but the performance critical Path, FileSystemPath and ScenePath are now implemented directly in C++. I've also replaced the old `Path.info()` method which returned a dictionary with a new pair of `Path::attributeNames()` and `Path::attribute()` methods. This allows a single attribute to be queried at a time (a requirement Ivan had a long while back) and works around the fact that we don't have a direct equivalent for dict in pure C++. Attributes must be derived from RunTimeTyped, which is the most flexible we can be within the Cortex framework, but is still more restrictive than the pure Python implementation. The old `info()` method isn't visible in C++ at all, but I've emulated it in the bindings to keep old Python code working, and also provided default implementations for the new attribute API that use the old `info()` method if its available. I hope this will ease the transition between the two implementations.

The PathListingWidget rewrite means that we must now link to Qt from GafferUI, and there are some details to be aware of there and which are documented in the comments for 6f4e139. I'll write a separate internal mail about the necessary environment changes at Image Engine. PathListingWidget columns are now implemented in C++ too, which means they have lost a little flexibility compared to their old Python counterparts, but they have much better performance. 808b1c0 and cbb6a3d demonstrate how to update code to deal with that.

I don't think this is quite there yet, and I'm still waiting to see if the Travis tests are passing perfectly, but I wanted to get something up for discussion. In particular, I'd really like feedback on the following :

- Is "attribute" a good name for "something you can query about a Path"? Are we worried about potential confusion with scene attributes? I didn't want to overload `info()` to avoid confusion with the old method, and also `infoNames()` sounds awful. I thought about "property", but that's got overlap with Qt. Any other ideas, or are you happy with attribute?
- Many of the Path implementations which remain in Python are still implementing `info()` and relying on the emulation layer in the bindings to implement the attribute methods. That was useful to me in testing, but should I go through and replace them with the proper implementation. Or replace all but one, so we still have one testing the emulation?

johnhaddon commented:
I've fixed a bug and uploaded an updated set of dependencies with cortex 9.0.0-b3 and shiboken included, so the Travis tests now pass.

andrewkaufman commented:
I haven't reviewed the code yet, but any reason you included the version bump here rather than waiting until it was merged?

carstenkolve commented:
- Is "attribute"/"property" a good name for "something you can query about a Path"?

"trait", "feature"?

johnhaddon commented:
This branch requires Cortex 9.0.0-b3, and updated dependencies packages which include the Qt headers and the Shiboken Python module. I needed to update the SConstruct and IE config in order to be able to build that package and upload it so that the Travis tests could pass. As part of that I also needed to bump the version number to give the package it's proper version. In the past I've done this without committing the changes, and ended up committing them too late, and having no true record of how the dependencies were built. So this time I thought it was best to have a commit where I could say "that's how the dependencies were built".

Really, I think we need to move the dependencies build logic into the gafferDependencies project so that it's a standalone thing, but I haven't found them time. My current plan is to do it when I get GafferAppleseed integrated into the default builds, as that will require additions to the dependencies build anyway.

johnhaddon commented:
Traits and features both sound a bit on/off to me - like a path has them or it doesn't, rather than all paths have them, but with potentially varying values. Does that make sense?

andrewkaufman commented:
> Many of the Path implementations which remain in Python are still implementing info() and relying on the emulation layer in the bindings to implement the attribute methods. That was useful to me in testing, but should I go through and replace them with the proper implementation. Or replace all but one, so we still have one testing the emulation?

Lets go ahead and replace them. Better to have all the examples showing the thing we want people doing. Maybe we can have a manufactured example in the tests to verify the `info()` approach?

andrewkaufman commented:
I get this in my terminal when using the Ops mode of iebrowser:

```
WARNING : PathListingWidget : TypeError: No registered converter was able to produce a C++ rvalue of type boost::intrusive_ptr<IECore::RunTimeTyped const> from this Python object of type list
```

I presume that's something in ieGafferExtensions that needs fixing up?

andrewkaufman commented:
> Is "attribute" a good name for "something you can query about a Path"? Are we worried about potential confusion with scene attributes? I didn't want to overload info() to avoid confusion with the old method, and also infoNames() sounds awful. I thought about "property", but that's got overlap with Qt. Any other ideas, or are you happy with attribute?

I am a little worried about the confusion with scene attributes... it'll be pretty hard to explain to a user... "There's an issue with the attributes on the scene path. No, not the attributes on the path in the scene, the attributes on the thing that lets you look at the path in the scene".

@paulondc didn't mind `properites`, he didn't think people would be confused with Qt properties.
@ivanimanishi said that Jabuka uses `infoKeys` currently, and he doesn't mind that. Or if you really don't like having info in there, maybe just `keys()`?

johnhaddon commented:
I've added a bunch of commits to address your comments. Things still outstanding :

- Prefix the exception for out of range exceptions? I guess I don't mind either way very much, but what's there matches what was in Python, and I never had need for more. I guess in C++ it can be handy to have a prefix, but in Python you get that from the stacktrace anyway.
- "Looks funny". Not a very compelling argument, but I can change it if you really want.
- Remove info() from python implementations and replace with attributes() and attributeNames(). Agreed - I'd like to get agreement on the right names for attributes before I do this though, so I don't have to do it twice.
-  `'Path' object has no attribute '_setCaches'`. You mailed offline about this and I've replied offline too, I'm just putting it here so we have a complete list of all outstanding questions.

That just leaves the question of the naming. I decided to move away from `info()` for a few reasons :

- It started to sound funny. I guess it sounds like a plural thing to me, so getting one at a time of it sounded off. This is the least compelling reason.
- Jabuka has overloaded it to optionally take an infoKey, and I really didn't want to try to translate that into two C++ virtual functions with the same name, different return types (one dict-like one and one RunTimeTyped one), and then try to manage compatibility in the bindings as well.
- We don't even have a dictionary-like type in C++ that can hold generic RunTimeTyped objects - the closest we have is CompoundData or CompoundObject. I felt it was important to support RunTimeTyped to keep it as flexible as possible, and it seemed that the one-at-a-time query was what we really wanted anyway (Ivan made a very good case for it in the past). So a new API plus a backwards compatibility layer for the short term seemed the best way.

That leaves us with properties vs attributes I think. I find it hard to imagine the scenario where we're trying to explain Path attributes vs ScenePlug attributes to a user - only the latter are really user facing in any way. So I think it's more down to what us programmers think. I agree that the clash with Qt properties isn't really a concern - I guess it felt like one to me because I was working on Qt/Paths code at the time, but we explicitly exclude Qt from the public API, so it's not really relevant from an outside perspective.

I don't feel crazily strongly either way, and I'm a bit biased by being the poor typist in this case. Maybe properties would be better? Better enough to warrant changing it? Wanna have a quick chat with the guys to get a consensus and then I'll go with the majority?

andrewkaufman commented:
> Prefix the exception for out of range exceptions

Yeah, I was thinking about the lack of full stacktrace making it more relevant.

> properties vs attributes

Consensus is to go with `properties`.

johnhaddon commented:
OK, I've replaced attributes with properties (and I think I prefer it), and updated all subclasses to implement the property API rather than the old info one. Good to go?

andrewkaufman commented:
Code looks good, but I'm marking this as pr-externalDependency until we do some testing with users. I'm hoping to get @chelmer and Norrie on it today.

johnhaddon commented:
How about we merge anyway, so I can work on top of it and get a bit more testing done? If we need to push out anything else in a hurry we can always use a 0.9_maintenance branch to do so...

johnhaddon commented:
I've pushed a fix for a crash which could occur if a plug was removed from its node, while being observed by a ScenePath which is then destroyed before the node is. I'm speculating that this will fix the crash that Ben observed in Caribou during testing, although I haven't reproduced his original problem.
